Add var support for recall preset

ploveman opened the pull request:
Add support for variable for recall preset setting.  Uses existing range of recall options to ensure the same values are allowed from the dropdown or variable.

ploveman commented:
@jswalden  at your earliest convenience; feedback very much welcome as my first attempt at typescript things.  Also option for any suggestions you might have for adding some tests (if you think warranted).

jswalden commented:
Hey, thanks for the work -- I've been busy investigating, then writing fixes and tests for regressions in v3.0.0, so haven't had a chance to think much beyond that.  Here are some tentative early thoughts at a _very_ cursory look.

I think probably there should just be a separate "Recall Preset by Expression" action (I don't know what name is best -- I suggest looking at other modules and seeing what names they use), rather than complicating up the existing one.  A checkbox _and_ `isVisible` use to control which arm gets invoked seems on first thought kinda herky-jerky.  :-)  It would be ideal if [`CompanionInputFieldDropdown`](https://bitfocus.github.io/companion-module-base/interfaces/CompanionInputFieldDropdown.html) supported `allowCustom` _and_ [`useVariables`](https://bitfocus.github.io/companion-module-base/interfaces/CompanionInputFieldTextInput.html)...but it doesn't.  (Technically `useVariables` I think only adds the "expressions supported" UI by the picker.  But I think that indicator is extremely desirable.)  So a separate action seems best.

I'm undecided as yet on tests.  Tests for UI can get a little bit too much you're just repeating the mechanism, sometimes.  That _might_ be the case for this.  I haven't looked closely enough yet at your changes.  If tests do stay, or appear in some form, test helper classes/functions should be in `__tests__` directories (as https://github.com/bitfocus/companion-module-ptzoptics-visca/blob/437e84b652c156972ecc01bd40063de00ad4e0fe/src/visca/__tests__/camera-interactions/run-test.ts does) so they get used/compiled for Jest stuff but aren't compiled into shipped module code.

On the TypeScript front, I see a bunch of `expr as Type` in this.  Generally it's better to put it in the variable declaration, I think, `const testEvent: CompanionActionEvent = { ... }`.

I'll try to get back with more in the next few days.

ploveman commented:
> Hey, thanks for the work -- I've been busy investigating, then writing fixes and tests for regressions in v3.0.0, so haven't had a chance to think much beyond that. Here are some tentative early thoughts at a _very_ cursory look.
> 
> I think probably there should just be a separate "Recall Preset by Expression" action (I don't know what name is best -- I suggest looking at other modules and seeing what names they use), rather than complicating up the existing one. A checkbox _and_ `isVisible` use to control which arm gets invoked seems on first thought kinda herky-jerky. :-) It would be ideal if [`CompanionInputFieldDropdown`](https://bitfocus.github.io/companion-module-base/interfaces/CompanionInputFieldDropdown.html) supported `allowCustom` _and_ [`useVariables`](https://bitfocus.github.io/companion-module-base/interfaces/CompanionInputFieldTextInput.html)...but it doesn't. (Technically `useVariables` I think only adds the "expressions supported" UI by the picker. But I think that indicator is extremely desirable.) So a separate action seems best.
> 
> I'm undecided as yet on tests. Tests for UI can get a little bit too much you're just repeating the mechanism, sometimes. That _might_ be the case for this. I haven't looked closely enough yet at your changes. If tests do stay, or appear in some form, test helper classes/functions should be in `__tests__` directories (as https://github.com/bitfocus/companion-module-ptzoptics-visca/blob/437e84b652c156972ecc01bd40063de00ad4e0fe/src/visca/__tests__/camera-interactions/run-test.ts does) so they get used/compiled for Jest stuff but aren't compiled into shipped module code.
> 
> On the TypeScript front, I see a bunch of `expr as Type` in this. Generally it's better to put it in the variable declaration, I think, `const testEvent: CompanionActionEvent = { ... }`.
> 
> I'll try to get back with more in the next few days.

Thanks for the feedback @jswalden  I modeled the use variable checkbox off an example from the x32 module, but happy to make it a separate action, I'll work on that over next few days.  Thanks for clarifying on the tests; I'll move them into a __tests__ folder if that seems ok.  The mock is a bit clunky, and maybe not worth it, but I'd prefer having the coverage if you think the test methodology seems reasonable.

ploveman commented:
@jswalden when you have a chance I think is ready for another review (I used From Variable because I couldn't find another module example and expressions don't seem supported from modules).  I couldn't figure out how to cast the string from the option to type string without an as.  There are some "as" in the tests, but I managed to change one out.  I don't know if it's equally  a concern in the mocking world or not, but felt ok for a cheap mock.  I'm also fine if you don't think the tests should be kept.  I like the fact they failed when I moved stuff around, but also not tied to them.

jswalden commented:
@ploveman One thing that's _very_ helpful with TypeScript: at least initially, use an IDE that displays type hints, offers autocomplete, etc.  [vscode](https://code.visualstudio.com/)'s been _super_ useful to me here, and in unfamiliar code -- plus it basically gets formatting right by default on save.  (There are some deviations I don't understand around line length, that I hope to figure out and eliminate eventually.)

To the other comments:

Regarding the look of the action/UI to the action, I think [this action](https://www.facebook.com/groups/companion/posts/3640774492807541/) shows a near-ideal UI, in every regard except for being so freeform you could enter any number (or non-number) you wanted and not get feedback on out-of-range or similar.  But if it'll accept variables, I guess that's the best you can do.

(One other advantage to explicitly signaling variable support that I neglected to mention: you get variable autocompletion when typing in the text input.)

Re mocks/tests for actions, I think my preference is for actions whose code is so short that you avoid bugs by it just being obvious.  And if you're testing what the action does, you separately test the idioms the action invokes.  Hence how `VISCAPort` gets tested but the actions that do stuff with it aren't.  (This is effectively "view" versus "model", if you're familiar with the model-view-controller idea.)

My understanding of TypeScript is that it's kind of deliberately designed as almost a preprocessor to JavaScript, and _you can break it_ if you actively try.  When you see `as` in code, that's not _doing_ anything: it's just a type cast for TypeScript's benefit, and it _trusts you_ to have actually ensured the type somehow.  If you want to actively convert, you need to use something like `String(val)` or `Boolean(val)`.  (I'll point these out in more detail when I review.

I still need to look at the actual code here in any depth.  :-)

jswalden commented on src/actions.ts:
```diff
@@ -333,6 +333,29 @@ export function getActions(instance: PtzOpticsInstance): CompanionActionDefiniti
+					return
+				}
+				event.options.val = hexval
+				void instance.sendCommand(PresetRecall, event.options)
```
You can just do

```typescript
    void instance.sendCommand(PresetRecall, { val: hexval })
```

rather than have to mutate/reuse the passed-in options.

jswalden commented on src/actions.ts:
```diff
@@ -333,6 +333,29 @@ export function getActions(instance: PtzOpticsInstance): CompanionActionDefiniti
+			],
+			callback: async (event: CompanionActionEvent) => {
+				const varPreset = await instance.parseVariablesInString(event.options.recallPresetVariableVal as string)
+				const hexval = PresetRecallOption.valmap.get(parseInt(varPreset, 10))
```
Add a

```typescript
export function twoDigitHex(n: number): string {
    return ('0' + n.toString(16)).slice(-2)
}
```

in `options.js` (and use it in the defining of `PRESET_CHOICES` so the same conversion technique is defined only one place), then do

```typescript
    const interpolated = await instance.parseVariablesInString(String(event.options.recallPresetVariableVal))
    const preset = parseInt(interpolated, 10)
    if (preset === NaN) {
        instance.log('error', `Invalid recall preset value: ${interpolated}`)
        return
    }
    void instance.sendCommand(PresetRecall, { val: twoDigitHex(preset) })
```

ploveman replied:
The only reason I used the map was to exclude that range in the middle that the drop-down doesn't support.  What do you think, similar to the first function, about having a common function that determines if the preset is in the "valid" range (excludes 90-100)?  That function could be called in the loop as well to exclude from the dropdown and to invalidate any variables?

ploveman replied:
I ended up adding an isValid to cover that as I mentioned

jswalden commented on src/camera/options.ts:
```diff
@@ -284,6 +287,8 @@ export const PresetSaveOption = {
 	id: 'val',
 	choices: PRESET_CHOICES,
+	valmap: PRESET_MAP,
+	useVariableId: 'useVariableForPreset',
```
Don't need this either with the action tweaks.

jswalden replied:
Unless Github UI is wrong, this in fact wasn't removed.

jswalden reviewed (changes requested):
I'm still on the fence about the mocks stuff. But _if_ it ends up removed, you might as well not remove it until the actions code is correct, having already written it.  I'll decide next go-around, I think.

ploveman commented:
@jswalden I think I've addressed all the review comments.  Thanks for the ideas.  The front-end tests worked well for me to verify that I didn't break anything with the refactor, but also understand if you don't want them around.

jswalden commented on src/actions.ts:
```diff
@@ -333,6 +335,29 @@ export function getActions(instance: PtzOpticsInstance): CompanionActionDefiniti
 			},
 		},
+		[PtzOpticsActionId.RecallPresetFromVar]: {
+			name: 'Recall Preset (by number)',
```
I hate to jerk you around like this.  😢  But:

By mostly coincidence I just found myself in a somewhat-similar situation: having an action whose effect is to send a command, and there's a bug with the command sent, and so I needed to test it.  Specifically, the user-specified parameters in the "Custom command" action weren't being filled right.

And it occurred to me: the way these slightly-complicated actions work is you have computations of some sort, and they lead up to a `void instance.sendCommand(...)`.

But `sendCommand` is merely the top end of the stack here.  Really, you want to verify that the conversion of command/options to bytes produces the right thing -- because if that works, `sendCommand` will too.  (And if you're really worried about `sendCommand` not sending those computed bytes, then you should be writing a `RunCameraInteractionTest` test -- and there are plenty of those.)

So the right way to formulate this sort of thing, is to move most of the body of this action into a standalone function: https://github.com/bitfocus/companion-module-ptzoptics-visca/commit/a269cbf3a4073448d922746b9fb17ebbff31fddb.  And then you test _that_: https://github.com/bitfocus/companion-module-ptzoptics-visca/commit/7dca219d235afd141c918f156201743ac0f22e32.

Amusingly, I happen to have created a mock context as part of this, that I just moved into its own standalone file.  https://github.com/bitfocus/companion-module-ptzoptics-visca/commit/d6a9396012510fc33233f61894a342d5cbbda271

So you should be able to move most of this into a standalone `export async function parsePresetVariableOption(option: string, context: CompanionActionContext): Promise<CompanionOptionValues>`.  And then you can create one ` MockContext` per test (don't create one and managing its state across tests: Jest supports running tests in isolation, and that sort of cross-test dependency breaks doing so), and in each test you can verify that `PresetRecall.toBytes(options)` has the expected value, as happens in one of those prior rev links.

ploveman replied:
@jswalden I'm happy to split this out, one concern I have though is that when I asked in slack about whether I should use context or instance for parsing variables, the answer was to use the instance vs the context.  I'm not sure what the implications are, but would prefer to follow that guidance.  Happy to take your context mock and merge it with some of the other instance mock needs (including log).  Thoughts?

jswalden commented on src/camera/options.ts:
```diff
@@ -265,10 +265,18 @@ export const AutoWhiteBalanceSensitivityOption = {
+	return ('0' + n.toString(16)).slice(-2)
+}
+
+export function isValidPreset(n: number): boolean {
```
Good call on breaking this out to a separate function and using it in the loop below here.

jswalden reviewed (changes requested):
Github has a "squash into one commit and rebase/merge that" mode for me to approve this eventually.  So I can mostly deal with having a dozen commits in one PR now: I can just use the "Files" view of the PR to see the combined effect.

But you might find your life easier if you squash them all together into a single commit (`git rebase --interactive master` while on your branch, then replace the various `pick` in your editor with `squash`).  And when you're modifying that one commit, after you `git add` each change, you use `git commit --amend` to "modify" that one commit (the old one's still around, just your branch won't point to it any more), rather than just keep accreting layers of tweaking atop it.  :smiley:

Again, this isn't the end of the world or anything.  I can deal if you don't do this.  But it will make for a slightly tidier display on Github for everyone if you do it.

ploveman commented:
@jswalden I think this is ready for review again, I think you may be able to squash when you merge.  I spent some time trying to look at this, but appears you can't squash after PR has been raised (or maybe you're asking for something different).

ploveman commented:
> @jswalden I think this is ready for review again, I think you may be able to squash when you merge. I spent some time trying to look at this, but appears you can't squash after PR has been raised (or maybe you're asking for something different).

@jswalden think this is ready when you have a moment

jswalden commented on src/__tests__/recall-by-number.test.ts:
```diff
@@ -0,0 +1,47 @@
+		const context = new MockContext()
+		const input: CompanionOptionValues = { val: '37' }
+		const result = await parsePresetVariableOption(input, context)
+		expect(typeof result).toBe('object')
```
Because `typeof null === 'object'`, I'm always leery of any comparison against `'object'`.  Moreover you're fighting the TypeScript type system in the very next line.  If you wrote slightly more idiomatic TypeScript that moved the type-check into control flow, then `result` would be automatically understood as a `CompanionOptionValues` in the next line.  I think you should do this as:

```ts
if (typeof result === 'string') {
    expect(typeof result !== 'string').toBe(true)
} else {
    const commandBytes = PresetRecall.toBytes(result)
    expect(prettyBytes(commandBytes)).toBe('[81 01 04 3F 02 25 FF]')
}
```

and similar for the other tests.

ploveman replied:
@jswalden  with your comment of refactoring the creation of the options into a function, what do you think about putting the verification into a function?  On one hand I like the idea for tidiness, but then errors come out of that function vs directly in the test and not sure if that clouds readability.  Welcome any thoughts you might have on that.

jswalden commented on src/__tests__/recall-by-number.test.ts:
```diff
@@ -0,0 +1,47 @@
+import { MockContext } from './mock-context.js'
+
+describe('test recall preset values', () => {
+	test('Variable value contant used dec 37', async () => {
```
Maybe `'User enters "37" as the preset'`, which I think correlates better with the UI operation being tested.

jswalden commented on src/__tests__/recall-by-number.test.ts:
```diff
@@ -0,0 +1,47 @@
+		expect(typeof result).toBe('object')
+		expect(prettyBytes(PresetRecall.toBytes(result as CompanionOptionValues))).toBe('[81 01 04 3F 02 25 FF]')
+	})
+	test('Variable value variable does not resolve to valid recall preset', async () => {
```
`'User enters "foo" (not a number at all) as the preset'`

jswalden commented on src/__tests__/recall-by-number.test.ts:
```diff
@@ -0,0 +1,47 @@
+describe('test recall preset values', () => {
+	test('Variable value contant used dec 37', async () => {
+		const context = new MockContext()
+		const input: CompanionOptionValues = { val: '37' }
```
Maybe add a global

```ts
function optionsWithRecallPresetOption(text: string): CompanionOptionValues {
    return { val: text }
}
```

and use that, to repeat the type and `val` only in one spot rather than all of them .

Also, can you rename `input` to `options`, seeing as that's what it's called everywhere else in code?  `input` confused me for a second there.

jswalden commented on src/__tests__/recall-by-number.test.ts:
```diff
@@ -0,0 +1,47 @@
+		expect(typeof result).toBe('string')
+		expect(result).toContain('Invalid recall preset value of')
+	})
+	test('Variable value constant invalid preset', async () => {
```
`'User enters a PTZOptics-reserved preset number'`

jswalden commented on src/__tests__/recall-by-number.test.ts:
```diff
@@ -0,0 +1,47 @@
+		expect(result).toContain('Invalid recall preset value of')
+	})
+
+	test('Variable value resolves to dec 254', async () => {
```
`'User enters a variable that resolves to a valid preset'`

jswalden commented on src/__tests__/recall-by-number.test.ts:
```diff
@@ -0,0 +1,47 @@
+		expect(prettyBytes(PresetRecall.toBytes(result as CompanionOptionValues))).toBe('[81 01 04 3F 02 FE FF]')
+	})
+
+	test('Invalid preset value dec 255', async () => {
```
`'User enters a variable that resolves to an invalid preset'`

jswalden commented on src/camera/options.ts:
```diff
@@ -265,10 +265,18 @@ export const AutoWhiteBalanceSensitivityOption = {
 }
 
+export function twoDigitHex(n: number): string {
+	return ('0' + n.toString(16)).slice(-2)
```
Let's change this to `n.toString(16).padStart(2, '0')` while we're here.

jswalden replied:
Looks like this is still undone.

jswalden commented on src/recall-by-number.ts:
```diff
@@ -0,0 +1,25 @@
+import { CompanionOptionValues } from '@companion-module/base'
+import { isValidPreset, twoDigitHex } from './camera/options.js'
+//import { PtzOpticsInstance } from './instance.js'
```
Remove this.

jswalden commented on src/actions.ts:
```diff
@@ -333,6 +334,27 @@ export function getActions(instance: PtzOpticsInstance): CompanionActionDefiniti
+				},
+			],
+			callback: async ({ options }, context) => {
+				const errorOrValue = await parsePresetVariableOption(options, context)
```
Name this `presetRecallOptions`.

ploveman replied:
Changed the ts file name to be generic since the same will apply equally to the save preset from number and left this parse function as is since will apply equally to both

jswalden reviewed (changes requested):
Sorry for the delay, I've been at work on another module whose code is in much gnarlier state and haven't been able to spare the time for this.

ChrisSW58 commented:
VERY interested in this when it arrives...well done! Question: will there be an equivalent to save a preset using a variable?

ploveman commented:
> VERY interested in this when it arrives...well done! Question: will there be an equivalent to save a preset using a variable?

This addition doesn't do a set from variable, but I think would be pretty easy to add once I get this in.  Thanks to @jswalden for his review in my learning typescript.  Will try and address latest comments this weekend sometime.

ploveman commented:
@jswalden  I think I've covered everything you asked for.  I ended up making a custom jest matcher since that seemed cleaner for the repeated tests.  I also changed the naming to just be preset vs recall-preset for the various functions since they can be re-used whenever we want to add save from preset.

ChrisSW58 commented:
Any news on how this is getting on?

ploveman commented:
> Any news on how this is getting on?

From the recall side, I think I've resolved the open concerns, but still need review from @jswalden .  Once we get this one merged, save from variable should be very straightforward

ploveman commented:
@jswalden just a reminder to please review.

ChrisSW58 commented:
Hi guys...is there any way this module - the recall module - will see the light of day anytime soon? Is there even a compiled beta version I could test with 3.4.0? It will make such a big difference to how we initialise our system so I'm waiting patiently. (Is there an escalation process needed to move it along in case there are issues not allowing to time on it?) Many thanks for your excellent work on this.

ploveman commented:
> Hi guys...is there any way this module - the recall module - will see the light of day anytime soon? Is there even a compiled beta version I could test with 3.4.0? It will make such a big difference to how we initialise our system so I'm waiting patiently. (Is there an escalation process needed to move it along in case there are issues not allowing to time on it?) Many thanks for your excellent work on this.

@ChrisSW58 unfortunately I don't have the permissions to merge; I believe I've responded to all of @jswalden comments, I don't know if anyone else has merge permissions to this module.  Asking @jswalden  for review again.  If this doesn't go through soon I think there may be a way to compile a beta of just the module for you, but it would have to be manually loaded.

ChrisSW58 commented:
> > Hi guys...is there any way this module - the recall module - will see the light of day anytime soon? Is there even a compiled beta version I could test with 3.4.0? It will make such a big difference to how we initialise our system so I'm waiting patiently. (Is there an escalation process needed to move it along in case there are issues not allowing to time on it?) Many thanks for your excellent work on this.
> 
> @ChrisSW58 unfortunately I don't have the permissions to merge; I believe I've responded to all of @jswalden comments, I don't know if anyone else has merge permissions to this module. Asking @jswalden for review again. If this doesn't go through soon I think there may be a way to compile a beta of just the module for you, but it would have to be manually loaded.

I'd appreciate that - it would be great if I can add it in simply by using the Companion Developers Module Path and pointing to it....

jswalden commented:
I've been working on the last bits of the allenheath-sq refactor (still). Might be able to finish that in the next few days. Unfortunately I'm traveling and the mobile data is...very weak, so we'll see what I can get done on this review after that. If I can't get enough usable data, it's going to be around August 12 that I return to civilization.

ChrisSW58 commented:
> I've been working on the last bits of the allenheath-sq refactor (still). Might be able to finish that in the next few days. Unfortunately I'm traveling and the mobile data is...very weak, so we'll see what I can get done on this review after that. If I can't get enough usable data, it's going to be around August 12 that I return to civilization.

Thank you Jeff - I would greatly appreciate that.

ChrisSW58 commented:
> I've been working on the last bits of the allenheath-sq refactor (still). Might be able to finish that in the next few days. Unfortunately I'm traveling and the mobile data is...very weak, so we'll see what I can get done on this review after that. If I can't get enough usable data, it's going to be around August 12 that I return to civilization.

As a matter of interest, is there any timescale which indicates when 3.4.4 will go live?

ploveman commented:
> > > Hi guys...is there any way this module - the recall module - will see the light of day anytime soon? Is there even a compiled beta version I could test with 3.4.0? It will make such a big difference to how we initialise our system so I'm waiting patiently. (Is there an escalation process needed to move it along in case there are issues not allowing to time on it?) Many thanks for your excellent work on this.
> > 
> > 
> > @ChrisSW58 unfortunately I don't have the permissions to merge; I believe I've responded to all of @jswalden comments, I don't know if anyone else has merge permissions to this module. Asking @jswalden for review again. If this doesn't go through soon I think there may be a way to compile a beta of just the module for you, but it would have to be manually loaded.
> 
> I'd appreciate that - it would be great if I can add it in simply by using the Companion Developers Module Path and pointing to it....

@ChrisSW58 I'm attaching a beta .tar.gz for this module that if you extract in a dev-modules dir and then point to that directory I was able to use the new function.  I created with the yarn companion-module-build and just renamed the dir so it was easier to recognize.  One caveat, I don't think there's anything that could go sideways from taking the beta of the module until it's in the companion beta/etc, but I also don't know if for 100% that it won't either.  Hope it helps tide you over until this gets merged.

[companion-module-ptzoptics-visca-recall-beta.tar.gz](https://github.com/user-attachments/files/16482995/companion-module-ptzoptics-visca-recall-beta.tar.gz)

ChrisSW58 commented:
> > > > Hi guys...is there any way this module - the recall module - will see the light of day anytime soon? Is there even a compiled beta version I could test with 3.4.0? It will make such a big difference to how we initialise our system so I'm waiting patiently. (Is there an escalation process needed to move it along in case there are issues not allowing to time on it?) Many thanks for your excellent work on this.
> > > 
> > > 
> > > @ChrisSW58 unfortunately I don't have the permissions to merge; I believe I've responded to all of @jswalden comments, I don't know if anyone else has merge permissions to this module. Asking @jswalden for review again. If this doesn't go through soon I think there may be a way to compile a beta of just the module for you, but it would have to be manually loaded.
> > 
> > 
> > I'd appreciate that - it would be great if I can add it in simply by using the Companion Developers Module Path and pointing to it....
> 
> @ChrisSW58 I'm attaching a beta .tar.gz for this module that if you extract in a dev-modules dir and then point to that directory I was able to use the new function. I created with the yarn companion-module-build and just renamed the dir so it was easier to recognize. One caveat, I don't think there's anything that could go sideways from taking the beta of the module until it's in the companion beta/etc, but I also don't know if for 100% that it won't either. Hope it helps tide you over until this gets merged.
> 
> [companion-module-ptzoptics-visca-recall-beta.tar.gz](https://github.com/user-attachments/files/16482995/companion-module-ptzoptics-visca-recall-beta.tar.gz)

Very many thanks 

> > > > Hi guys...is there any way this module - the recall module - will see the light of day anytime soon? Is there even a compiled beta version I could test with 3.4.0? It will make such a big difference to how we initialise our system so I'm waiting patiently. (Is there an escalation process needed to move it along in case there are issues not allowing to time on it?) Many thanks for your excellent work on this.
> > > 
> > > 
> > > @ChrisSW58 unfortunately I don't have the permissions to merge; I believe I've responded to all of @jswalden comments, I don't know if anyone else has merge permissions to this module. Asking @jswalden for review again. If this doesn't go through soon I think there may be a way to compile a beta of just the module for you, but it would have to be manually loaded.
> > 
> > 
> > I'd appreciate that - it would be great if I can add it in simply by using the Companion Developers Module Path and pointing to it....
> 
> @ChrisSW58 I'm attaching a beta .tar.gz for this module that if you extract in a dev-modules dir and then point to that directory I was able to use the new function. I created with the yarn companion-module-build and just renamed the dir so it was easier to recognize. One caveat, I don't think there's anything that could go sideways from taking the beta of the module until it's in the companion beta/etc, but I also don't know if for 100% that it won't either. Hope it helps tide you over until this gets merged.
> 
> [companion-module-ptzoptics-visca-recall-beta.tar.gz](https://github.com/user-attachments/files/16482995/companion-module-ptzoptics-visca-recall-beta.tar.gz)

Very many thanks - this allows me to start the code changes, knowing what the call now is. My only comment is that, for a new inexperienced user, the two actions - Recall Preset(by number) and Recall Preset are very similar when it comes to the input field - one has Preset Nr and the other has Preset Number (which also allows variables). Not sure why the two actions would not combine over time, but I'm delighted to have the ability to use variables. Very many thanks to both you and Jeff.

ploveman commented:
Thanks for the feedback @ChrisSW58 I actually started with a single action, but Jeff thought the UX wasn't good and the code is cleaner having it separate.  I think one of the cannon cameras did a similar separation.  That being said once this and the save are in I can try to see if there's a strategy that everyone is happy with (I'd leave the separate by number actions to avoid breaking folks) and then if we have both folks can chose their own adventure :)  I'm def not a UX person so front-end things are hard for me; but feedback from users like you are helpful

ChrisSW58 commented:
> Thanks for the feedback @ChrisSW58 I actually started with a single action, but Jeff thought the UX wasn't good and the code is cleaner having it separate. I think one of the cannon cameras did a similar separation. That being said once this and the save are in I can try to see if there's a strategy that everyone is happy with (I'd leave the separate by number actions to avoid breaking folks) and then if we have both folks can chose their own adventure :) I'm def not a UX person so front-end things are hard for me; but feedback from users like you are helpful

I think just changing the files name within the module would make it clearer...they are just too similar for people to deduce the difference if inexperience e.g. Preset Number/Variable would be a lot clearer. Just my two pennyworth! :) And I completely get the issue of clean code and code separation - ode things are written with no view to future maintenance.

ploveman commented:
I wonder if adding something like "Use number/variable" in the current drop-down might be interesting.  Maybe that'd be less jerky when hiding the text input.  The module code doesn't allow an option for disable which might be my pref vs hiding so it doesn't blip as much.

ChrisSW58 commented:
> I wonder if adding something like "Use number/variable" in the current drop-down might be interesting. Maybe that'd be less jerky when hiding the text input. The module code doesn't allow an option for disable which might be my pref vs hiding so it doesn't blip as much.

I’ll let you and Jeff as the experts sort it out. I’ve designed a number of successful UI systems but it’s really down to the two of you.

ploveman commented:
> Question - does the call support something like '$(internal:custom_Preset_Var1)-100'?

Unfortunately no, companion doesn't support expressions for module actions; which means you'd have to do the math in a separate step.

ChrisSW58 commented:
Congratulations - the system is working very well and exactly what I was wanting. Very many thanks.

ploveman commented:
@jswalden reminder this needs to be merged please

jswalden commented on src/__tests__/preset-by-number.test.ts:
```diff
@@ -0,0 +1,115 @@
+	expected_bytes
+) {
+	if (typeof actual === 'string') {
+		throw new TypeError('Input showuld be of type CompanionOptionValues not string')
```
"should"

jswalden commented on src/__tests__/preset-by-number.test.ts:
```diff
@@ -0,0 +1,115 @@
+	}
+
+	const commandBytes = command.toBytes(actual as CompanionOptionValues)
+	const pass = prettyBytes(commandBytes) === expected_bytes
```
Can't you just add `: CompanionOptionValues` next to `actual` in the function argument list?  We pretty much should never see `as` in actual code.

Comparing `prettyBytes` output is kinda skeevy compared to making the expected byte list an array of numbers.  But also it's test code, and sometimes I need to remember that it is okay for test code to be messier, as long as it's not unreadably messy.

ploveman replied:
Unfortunately no, it's not compatible with the matcher definition.  Otherwise would be happy to

jswalden replied:
I pulled the PR code and experimented with it -- `actual` is `unknown` here (and so you have to type-test it to do anything with it).  It's easy to test for a string.  But it's not actually trivial to use `actual` as a `CompanionOptionValues`.  I'm not even sure it's possible to do it without casting -- casting that is entirely avoidable if we just test the values directly, rather than defining a matcher.

At this point I think we're better off just eschewing a custom matcher: write normal, typed functions to do the tests, along lines something like:

```ts
function expectResolvesToCommandBytes(
    result: string | CompanionOptionValues,
    command: Command,
    expectedBytes: readonly number[],
): void {
    if (typeof result === 'string') {
        throw new TypeError('Result should not have been an error, instead got: ' + result)
    }

    const commandBytes = command.toBytes(result)
    expect(commandBytes.length).toBe(expectedBytes.length)
    for (let i = 0; i < commandBytes.length; i++) {
        expect(commandBytes[i]).toBe(expectedBytes[i])
    }
}

function expectIsErrorString(
    result: string | CompanionOptionValues,
): void {
    if (typeof result !== 'string') {
        throw new TypeError('Result should have been an error')
    }
}
```

And then have callers specify the command bytes as numbers, like `[0x81, 0x01, 0x04, 0x3F, 0x02, 0x25, 0xFF]`.

jswalden commented on src/__tests__/preset-by-number.test.ts:
```diff
@@ -0,0 +1,115 @@
+			`expected ${this.utils.printReceived(prettyBytes(commandBytes))} ${
+				pass ? 'not ' : ''
+			}to match\nreceived ${this.utils.printExpected(expected_bytes)}`,
+		pass: pass,
```
JavaScript/TypeScript allow you to just say "pass" as a shorthand to define a property with a name expression, with that name expression's value.

ploveman replied:
done

jswalden commented on src/__tests__/preset-by-number.test.ts:
```diff
@@ -0,0 +1,115 @@
+	}
+}
+
+declare module 'expect' {
```
I gotta be honest, I've been sitting and staring at this _mostly_ because I'm pretty leery about spinning up an entire custom matcher just for one function return type.  The closer we stick to relatively vanilla JS/TypeScript -- for example, writing helper functions to perform tests of all type-arms -- and the further from delving into test framework intricacies, the more understandable it all stays for Jest n00bs (like, um, me :innocent:).

But maybe I've stared at this enough to come around on it.

That said...this `declare module` goo is so much opaque "is it correct?  :shrug:" for me, that makes me hesitate...

ploveman replied:
From the jest examples for custom matcher  https://github.com/jestjs/jest/blob/bd1c6db7c15c23788ca3e09c919138e48dd3b28a/examples/expect-extend/toBeWithinRange.ts

jswalden reviewed (changes requested):
I see we're closing in on another release cycle going live -- let's make sure this goes in for it.

ploveman commented:
@jswalden done, made the changes I can and followed jest example for extension (figure they know best).

ploveman commented:
@jswalden I don't have time to refactor tests again.  As you said they're tests, more tests than most things have.  If you want to refactor feel free.

jswalden commented:
Fair enough.  I'll make the tweaks I want and land it.

ChrisSW58 commented:
Gents - my heartfelt thanks to the both of you for your hard work.

jswalden commented:
Landed as e13984545650a7c5e46efdc604bd23b96e55145f.

jswalden commented:
I spun https://github.com/bitfocus/companion-module-ptzoptics-visca/releases/tag/v3.1.0-rc1 as a prerelease for anyone who wants to test it.  I haven't bumped Companion dependencies because there hasn't been particular reason to, so this should work in Companion 3.2 or later.

ChrisSW58 commented:
Gents, with the success in delivering this ability to recall a preset with a variable, how and where can I raise a request to develop the other half of the coin - the ability to save a preset with a variable?

ploveman commented:
@ChrisSW58  I may try and knock out a quick PR tomorrow, with the other pieces in place from the recall, believe should be just a few changes to add the action and tests for it.

ploveman commented:
@ChrisSW58  PR for save up here https://github.com/bitfocus/companion-module-ptzoptics-visca/pull/60
